addons/stock/models/stock_picking.py

```diff
@@ -948,44 +948,55 @@ def _check_entire_pack(self):
                         for pl in package_level_ids:
                             pl.location_dest_id = self._get_entire_pack_location_dest(pl.move_line_ids) or picking.location_dest_id.id
 
-    def do_unreserve(self):
-        self.move_ids._do_unreserve()
-        self.package_level_ids.filtered(lambda p: not p.move_ids).unlink()
+    def _get_lot_move_lines_for_sanity_check(self, none_done_picking_ids, separate_pickings=True):
+        """ Get all move_lines with tracked products that need to be checked over in the sanity check.
+            :param none_done_picking_ids: Set of all pickings ids that have no qty_done set on any move_line.
+            :param separate_pickings: Indicates if pickings should be checked independently for lot/serial numbers or not.
+        """
+        def get_relevant_move_line_ids(none_done_picking_ids, picking):
+            # Get all move_lines if picking has no qty_done set, otherwise only get the move_lines with some qty_done set.
+            return picking.move_line_ids.ids if picking.id in none_done_picking_ids else get_line_with_done_qty_ids(picking.move_line_ids)
 
-    def button_validate(self):
-        # Clean-up the context key at validation to avoid forcing the creation of immediate
-        # transfers.
-        ctx = dict(self.env.context)
-        ctx.pop('default_immediate_transfer', None)
-        self = self.with_context(ctx)
+        def get_line_with_done_qty_ids(move_lines):
+            # Get only move_lines that has some qty_done set.
+            return [line.id for line in move_lines if float_compare(line.qty_done, 0, precision_rounding=line.product_uom_id.rounding)]
 
-        # Sanity checks.
-        pickings_without_moves = self.browse()
-        pickings_without_quantities = self.browse()
+        if separate_pickings:
+            # If pickings are checked independently, get full/partial move_lines depending if each picking has no qty_done set.
+            lines_to_check_ids = [line_id for picking in self for line_id in get_relevant_move_line_ids(none_done_picking_ids, picking)]
+        else:
+            # If pickings are checked as one (like in a batch), then get only the move_lines with qty_done across all pickings if there is at least one.
+            lines_to_check_ids = self.move_line_ids.ids
+            if any(picking.id not in none_done_picking_ids for picking in self):
+                lines_to_check_ids = get_line_with_done_qty_ids(self.move_line_ids)
+
+        return self.env['stock.move.line'].browse(lines_to_check_ids).filtered(lambda ml: ml.product_id and ml.product_id.tracking != 'none')
```

Comment:
instead of looping through all the move lines more times than necessary, can we do something like:

```
def get_relevant_move_line_ids(none_done_picking_ids, picking):
    # Get all move_lines if picking has no qty_done set, otherwise only get the move_lines with some qty_done set.
    if picking.id in none_done_picking_ids:
        return picking.move_line_ids.filtered(lambda ml: ml.product_id and ml.product_id.tracking != 'none').ids
    else:
        return get_line_with_done_qty_ids(picking.move_line_ids)

def get_line_with_done_qty_ids(move_lines):
    # Get only move_lines that has some qty_done set.
    return move_lines.filtered(lambda ml: ml.product_id and ml.product_id.tracking != 'none' and float_compare(ml.qty_done, 0, precision_rounding=ml.product_uom_id.rounding)).ids
```

Reply:
Agreed.
Didn't put it this way because I felt it unnecessary to do the filter at three separate places when it could be done once in the end. But I didn't take perfs into consideration there I admit.